### PR TITLE
Make linked samples parseable

### PR DESCRIPTION
**Pull request by gspencergoog**
## Description

This modifies the sample parser to be able to understand samples that are "linked" by a reference in them, and read those samples off disk instead of trying to parse them from the sample itself.

This will enable the sample analyzer and dartdoc to be able to generate correct output.

### CODE REVIEW

File: .run/Extract.run.xml [hunk 1]
```diff
@@ -6,8 +6,6 @@
     </option>
     <option name="filePath" value="$PROJECT_DIR$/packages/snippets/bin/extract.dart" />
     <option name="workingDirectory" value="$PROJECT_DIR$/../flutter" />
-    <method v="2">
-      <option name="ToolBeforeRunTask" enabled="true" actionId="Tool_External Tools_Remove API" />
-    </method>
+    <method v="2" />
   </configuration>
 </component>
```
Comment (darrenaustin) [inline .run/Extract.run.xml:11]: I'm not familiar with these .run files. What are they configuring? Also should they end with a new line?

File: packages/snippets/bin/extract.dart [hunk 2]
```diff
@@ -77,65 +77,77 @@ Future<void> main(List<String> argList) async {
...
     }
+  } on SnippetException catch (e, s) {
+    print('Failed: $e\n$s');
+    exit(1);
+  } on FileSystemException catch (e, s) {
+    print('Failed with file system exception: $e\n$s');
+    exit(2);
+  } catch (e, s) {
+    print('Failed with exception: $e\n$s');
+    exit(2);
```
Comment (darrenaustin) [inline packages/snippets/bin/extract.dart:150]: Any significance to the 2 here, or is just another positive non-zero number?

File: packages/snippets/bin/snippets.dart [hunk 3]
```diff
@@ -96,6 +96,7 @@ String getChannelNameWithRetries() {
 void main(List<String> argList) {
   const Platform platform = LocalPlatform();
   final Map<String, String> environment = platform.environment;
+
```
Comment (darrenaustin) [inline packages/snippets/bin/snippets.dart:99]: Minor nit: space here needed?

File: packages/snippets/test/snippet_parser_test.dart [hunk 4]
```diff
@@ -211,11 +247,51 @@ class DocumentedClass {
...
+  /// String elementName = 'DocumentedClass.member';
   /// ```
   /// {@end-tool}
   void member() {}  
 }
 ''');
 }
+
+File _createDartpadSourceFile(Directory tmpDir, FileSystem filesystem, Directory flutterRoot, {bool linked = false}) {
+  final File linkedFile = filesystem.file(path.join(flutterRoot.absolute.path, 'linked_file.dart'))..createSync(recursive: true)..writeAsStringSync('''
```
Comment (darrenaustin) [inline packages/snippets/test/snippet_parser_test.dart:259]: Minor nit: might want to break up this line as it is a little hard to read in one chunk.

**Review by darrenaustin — APPROVED** [review]
LGTM. Just some minor comments and questions.

**Review by goderbauer — APPROVED** [review]
LGTM

File: .run/Extract.run.xml (hunk 1 above)
Comment (gspencergoog) [reply .run/Extract.run.xml:11]: They're for IntelliJ, so you can share run configurations. They don't have to end with a newline (that's how IntelliJ writes them).

File: packages/snippets/bin/extract.dart (hunk 2 above)
Comment (gspencergoog) [reply packages/snippets/bin/extract.dart:150]: Just another positive, non-zero number, and 2 to differentiate it from a "snippet exception" which uses `exit(1)`, although I don't really do that anywhere.

File: packages/snippets/bin/snippets.dart (hunk 3 above)
Comment (gspencergoog) [reply packages/snippets/bin/snippets.dart]: Yeah, that's random.  Removed.

File: packages/snippets/test/snippet_parser_test.dart (hunk 4 above)
Comment (gspencergoog) [reply packages/snippets/test/snippet_parser_test.dart]: Sure, done.

File: packages/snippets/CHANGELOG.md [hunk 5]
```diff
@@ -2,6 +2,10 @@
 
 ## 0.2.1
 
+* Enable parsing of linked samples, add section markers to extracted samples.
+
+## 0.2.1
```
[the remark this answers is not in the document]
Comment (gspencergoog) [reply packages/snippets/CHANGELOG.md:7]: Line 3 should be bumped, sorry.

File: packages/snippets/lib/data_types.dart [hunk 6]
```diff
@@ -4,6 +4,7 @@
 
 import 'package:args/args.dart';
 import 'package:file/file.dart';
+import 'package:snippets/snippets.dart';
```
[the remark this answers is not in the document]
Comment (gspencergoog) [reply packages/snippets/lib/data_types.dart]: Fixed.

**Pull request merged**